p2p/switch.go

```diff
@@ -369,21 +369,44 @@ func (sw *Switch) IsDialing(addr *NetAddress) bool {
 
 // Broadcast runs a go routine for each attempted send, which will block
 // trying to send for defaultSendTimeoutSeconds. Returns a channel
-// which receives success values for each attempted send (false if times out).
+// which receives broadcast result for each attempted send (success=false if times out).
 // NOTE: Broadcast uses goroutines, so order of broadcast may not be preserved.
 // TODO: Something more intelligent.
-func (sw *Switch) Broadcast(chID byte, msg interface{}) chan bool {
-	successChan := make(chan bool, len(sw.peers.List()))
+
+type BroadcastResult struct {
+	PeerKey string
+	Success bool
+}
+
+func (sw *Switch) Broadcast(chID byte, msg interface{}) chan BroadcastResult {
+	successChan := make(chan BroadcastResult, len(sw.peers.List()))
 	sw.Logger.Debug("Broadcast", "channel", chID, "msg", msg)
 	for _, peer := range sw.peers.List() {
 		go func(peer Peer) {
 			success := peer.Send(chID, msg)
-			successChan <- success
+			successChan <- BroadcastResult{peer.Key(), success}
 		}(peer)
 	}
 	return successChan
 }
 
+func (sw *Switch) TryBroadcast(chID byte, msg interface{}) chan BroadcastResult {
+	successChan := make(chan BroadcastResult, len(sw.peers.List()))
+	sw.Logger.Debug("TryBroadcast", "channel", chID, "msg", msg)
+	for _, peer := range sw.peers.List() {
+		success := peer.TrySend(chID, msg)
+		if success {
+			successChan <- BroadcastResult{peer.Key(), success}
+		} else {
+			go func(peer Peer) {
+				success := peer.Send(chID, msg)
```

Comment:
so this could potentially block forever?

Reply:
oh whoops, where does this block? trySend is non-blocking, no? and successChan is buffered

Reply:
peer.Send is blocking

Reply:
right, so that will block for defaultSendTimeoutSeconds?

Reply:
If TrySend were to succeed, Send would also succeed immediately, so if one-shot-nonblocking goroutines are sufficiently cheap, this function provides no benefit.  Lets do a performance benchmark and see what our options are.

The naming is inconsistent.  Its behavior is basically identical to Broadcast, so there's nothing more "try"-y about this.

Reply:
should we consider trying https://github.com/Jeffail/tunny to limit number of goroutines? or it's better to write our own goroutine pool?

Reply:
seems reasonable. It's small enough that we could probably audit it in less time than it would take to re-write the functionality